include/swift/Basic/ExperimentalDependencies.h

```diff
@@ -0,0 +1,968 @@
+/// quo), by a pair of nodes. One node represents the "implementation." If the
+/// implementation changes, users of the entity need not be recompiled. The
+/// other node represents the "interface." If the interface changes, any uses of
+/// that definition will need to be recompiled. The implementation always
+/// depends on the interface, since any change that alters the interface will
+/// require the implementation to be rebuilt. The interface does not depend on
+/// the implementation. In the dot files, interfaces are yellow and
+/// implementations white. Each node holds an instance variable describing which
+/// aspect of the entity it represents.
+enum class DeclAspect { interface, implementation, aspectCount };
+const std::string DeclAspectNames[]{"interface", "implementation"};
+
+template <typename FnT> void forEachAspect(FnT fn) {
+  for (size_t i = 0; i < size_t(DeclAspect::aspectCount); ++i)
+    fn(DeclAspect(i));
+}
+
+/// A pair of nodes that represent the two aspects of a given entity.
+/// Templated in order to serve for either FrontendNodes or DriverNodes.
+template <typename NodeT> class InterfaceAndImplementationPair {
+  NodeT *interface;
+  NodeT *implementation;
+
+public:
+  InterfaceAndImplementationPair()
+      : interface(nullptr), implementation(nullptr) {}
+
+  InterfaceAndImplementationPair(NodeT *interface, NodeT *implementation)
+      : interface(interface), implementation(implementation) {
+    assert(
+        interface->getKey().isInterface() &&
+        implementation->getKey().isImplementation() &&
+        "Interface must be interface, implementation must be implementation.");
+  }
+
+  NodeT *getInterface() const { return interface; }
+  NodeT *getImplementation() const { return implementation; }
+
+  /// When creating an arc to represent a link from def to use, the use end of
+  /// the arc depends on if the dependency is a cascading one. Centralize that
+  /// choice here.
+  /// ("use" in the name represents the noun, not the verb.)
+  NodeT *useDependingOnCascading(bool ifCascades) {
+    return ifCascades ? interface : implementation;
+  }
+};
+
+//==============================================================================
+// MARK: DependencyKey
+//==============================================================================
+
+/// The dependency system loses some precision by lumping entities together for
+/// the sake of simplicity. In the future, it might be finer-grained. The
+/// DependencyKey carries the information needed to find all uses from a def
+/// because the data structures in the graph map the key of an entity to all the
+/// nodes representing uses of that entity, even though the node may not really
+/// use the entity. For example, argument names of functions are ignored.
+class DependencyKey {
+  NodeKind kind;
+  DeclAspect aspect;
+  /// The mangled context type name of the holder for \ref potentialMember, \ref
+  /// member, and \nominal kinds. Otherwise unused.
```

Comment:
Missing `\ref` here.

Reply:
Arguably a "nominal" node has a name but no context, but I guess it doesn't really matter.

Reply:
Thanks. Added the \ref. 

I went back-and-forth on which field to use for the "nominal" node. Could go either way, and `DependencyKey` encapsulates the choice. Since it uses `mangleTypeAsContext` (as the current system does), I ended up with a context but no name. If you think there's a good reason, I can switch it.